NEXT-12691 - Add message to queue (#174)
shopware/docs: guides/plugins/plugins/framework/message-queue/add-message-to-queue.md

# Add message to queue

> ## Overview

Reviewer: I'm missing a general explanation, what a the message queue is. 

The message queue allows you to ... 

Maybe you can copy something from the symfony docs as some sort of introduction here

In this guide you'll learn how to create a message and add it to the queue.

> A [message](https://symfony.com/doc/current/messenger.html#creating-a-message-handler) is a simple PHP class that you want to dispatch over the MessageQueue. 

Reviewer: I'd prefer the terminology `object`.

```suggestion
A [message](https://symfony.com/doc/current/messenger.html#creating-a-message-handler) is a simple PHP object that you want to dispatch over the MessageQueue. 
```

class SmsNotification
{
    private $content;

    public function __construct(string $content)
    {
        $this->content = $content;
    }

    public function getContent(): string
    {
        return $this->content;
    }
}
```
{% endcode %}

## Send a message

After we've created our notification, we will create a service that will send our `SmsNotifcation`.
We will name this service `ExampleSender`. In this service we need to inject the `Symfony\Component\Messenger\MessageBusInterface`,
that is needed to send the message through the desired bus, which is called `messenger.bus.shopware`.

{% code title="<plugin root>/src/Service/ExampleSender.php" %}
```php
<?php declare(strict_types=1);

namespace Swag\BasicExample\Service;

use Swag\BasicExample\MessageQueue\Message\SmsNotification;
use Symfony\Component\Messenger\MessageBusInterface;

class ExampleSender
{
    /**
     * @var MessageBusInterface
     */
    private $bus;

    public function __construct(MessageBusInterface $bus)
    {
        $this->bus = $bus;
    }

    public function sendMessage(string $message): void
    {
        $this->bus->dispatch(new SmsNotification($message));
    }
}
```
{% endcode %}

> If we want to add metadata to our message, we can simply dispatch an `Symfony\Component\Messenger\Envelope` with the necessary 

Reviewer: ```suggestion
If we want to add metadata to our message, we can dispatch an `Symfony\Component\Messenger\Envelope` in our Service instead with the necessary 
```

[stamps](https://symfony.com/doc/current/components/messenger.html#adding-metadata-to-messages-envelopes).
In this example below, we use the `Symfony\Component\Messenger\Stamp\DelayStamp`, which tells the queue to process the
message later.

{% code title="<plugin root>/src/Service/ExampleSender.php" %}
```php
public function sendMessage(string $message): void
{
    $message = new SmsNotification($message);

>     $this->bus->dispatch(
>         (new Envelope($message))
>         ->with(new DelayStamp(5000))
>     );

Reviewer: ```suggestion
    $this->bus->dispatch(
        (new Envelope($message))
            ->with(new DelayStamp(5000))
    );
```
